### PR TITLE
Add .envrc(direnv) support

### LINKED ISSUE #102: Request to add support for .envrc format
We are using https://direnv.net/ to share a base config and it works really well for that. Would be even better if we could reference the same file in our run configs.

Format is pretty simple:

export DB_NAME=abc
export DB_HOSTNAME=www.hosts.com
export DB_PORT=3333

Thanks!

### PR DESCRIPTION
**Pull request by shinichy**
Fixes #102

* Add `DirenvProvider` to support `.envrc` file
* Add `getFileNamePredicate` and `showHiddenFiles` methods to `EnvVarsProviderFactory` interface to allow selecting only `.envrc` file.

### CODE REVIEW

**Comment by shinichy** [comment]
@ashald Would you take a look?

**Comment by cmmoran** [comment]
Incidentally, you don't need to add `direnv` to `/usr/local/bin` or any other "pointy" location. Creating a `.zprofile` (for zsh for example) with proper PATH setup will enable IntelliJ (or any other IDE or Spotlight/Dock launched app from macOS) to obtain a much more usable set of environment variables. 

**See: [MacOS - ZSH: .zprofile, .zshrc, .zlogin...](https://itectec.com/askdifferent/macos-zsh-zprofile-zshrc-zlogin-what-goes-where/)**

File: README.md [hunk 1]
```diff
@@ -226,7 +227,13 @@ In certain cases (so far, only `Gradle` has been confirmed) the implementation e
 but doesn't provide interfaces for it to actually make its work. Luckily, it was possible to make few assumptions about
 IDEA's internal implementation and make it work. Such integration is very fragile and it's not immediately clear if it
 will affect any existing integrations and when it will break. For that reason there is a special option to
-`Enable Experimental Integrations` that can be enabled when desired and should prevent other integrations from breaking. 
+`Enable Experimental Integrations` that can be enabled when desired and should prevent other integrations from breaking.
+
+#### direnv
+You need `direnv` in your `PATH` environment variable. `direnv allow` is run automatically if needed.
+
+Note: On macOS, you may need to add `/usr/local/bin` to the system-wide `PATH` using `launchctl` because `PATH` is not set correctly when IntelliJ is launched from Dock or Spotlight.
```
Comment (shinichy) [inline README.md]: @cmmoran Thanks for the tips! Do you think this note is not necessary?

Comment (cmmoran) [reply README.md]: Correct. As long as you have a `.zprofile` (zsh) or `.profile` (bash) with `direnv` in the PATH, then the JetBrains toolbox of apps will pick up the PATH and `direnv` will operate properly.

Comment (shinichy) [reply README.md]: I see. Thanks!

**Comment by shinichy** [comment]
@ashald Do you have a chance to review this PR?

**Comment by ashald** [comment]
I'd be glad to merge `.envrc` support, thought this PR breaks abstraction where `core` module is agnostic of intelij. I might get some time to look into fixing it in November, but if you could update it accordingly it'd help get it merged sooner.

**Comment by shinichy** [comment]
@ashald Thank you for reviewing this PR. I updated this PR not to break the core module abstraction. Would you take a look again?

**Comment by ashald** [comment]
@shinichy ugh, my apologies, I just worked on refactoring some mess in core module to make the interface cleaner. I just submitted a release, and unless there is a critical bug, I don't plan to make any changes to the code. Could you please rebase your changes one more? Apologies for back and forth.

**Comment by shinichy** [comment]
@ashald No worries. I fixed all the conflicts. Could you please take a look again?

File: modules/core/src/main/java/net/ashald/envfile/providers/direnv/DirenvProvider.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        } catch (Exception ex) {
+            throw new EnvFileException(ex);
+        }
+    }
+
+    private Process executeDirenv(String workingDir, String... args) throws IOException {
+        String[] newArgs = new String[args.length + 1];
+        System.arraycopy(args, 0, newArgs, 1, args.length);
+        newArgs[0] = "direnv";
```
Comment (RockyMM) [inline modules/core/src/main/java/net/ashald/envfile/providers/direnv/DirenvProvider.java:64]: I might be reading this incorrectly, but is this relying on an external tool? If yes, I would document this somewhere, maybe even in `README.md`.

Comment (RockyMM) [reply modules/core/src/main/java/net/ashald/envfile/providers/direnv/DirenvProvider.java:64]: Ah, sorry, I was not aware of this tool :)

**Review by RockyMM — APPROVED** [review]

File: modules/core/src/main/java/net/ashald/envfile/providers/direnv/DirenvProvider.java (hunk 2 above)
Comment (shinichy) [reply modules/core/src/main/java/net/ashald/envfile/providers/direnv/DirenvProvider.java:64]: Yes, I added a document about `direnv` tool in `README.md` originally, but deleted it by accident. I added it again.

Comment (shinichy) [reply modules/core/src/main/java/net/ashald/envfile/providers/direnv/DirenvProvider.java:64]: Thanks for catching it!

**Comment by ashald** [comment]
Hi @shinichy, apologies for a delayed reply.
I was thinking about it for a long time and eventually I realized I would rather not merge in providers that depend on external tools.

What I can offer to you instead is the extension point interface such that you could pack your changes as an independent plugin that relies on extension points provided by EnvFile, but can be installed separately. Please let me know if anything is missing in the current interface that would be necessary for you to implement the above.

P.S.: Alternatively, the latest version of EnvFile can execute any given file and interpret stdout as .env or JSON/YAML. To the best of my understanding, you could wrap a call to direnv in a tiny script that would just output env vars you set.

**Comment by shinichy** [comment]
@ashald Thanks for letting me know. I don't use IntelliJ and direnv these days, so I'll check your suggestions when I have a chance to use IntelliJ again.

**Pull request closed without merging**